### PR TITLE
Feat/api service

### PR DESCRIPTION
**Pull request by KirillSerg**
- create src/services/api.ts
- install [axios](https://www.npmjs.com/package/axios)
- create instance of axios with base URL that gets from
- research and add interceptors to manage token

### CODE REVIEW

**Review by Mykyta-Pysarev — APPROVED** [review]
Looks complicated. Hope it works :)

There is a lot of comments in the code. Do we really need them?

File: src/services/api.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  return localStorage.getItem("refreshToken");
+};
+
+const updateLocalAccessToken = (accessToken: string): void => {
+  localStorage.setItem("accessToken", accessToken);
+};
+
+const instance = axios.create({
+  baseURL: "http://localhost:3001/api", // /auth/login[,register]
+  headers: {
```
Comment (VitaliiDenysiuk) [inline src/services/api.ts:20]: - do not save url inside module, we can use different url for production and develop.
move "http://localhost:3001/" in .env
- add .env in .gitignore
- use .env functionality

[examle .env functionality](https://stackoverflow.com/questions/49579028/adding-an-env-file-to-react-project) for this

remove part "/api" from baseURL. Some our url will have not such part

File: src/services/api.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+instance.interceptors.request.use(
+  (config) => {
+    const token = getLocalAccessToken();
+    if (token) {
+      // config.headers.Authorization = "Bearer " + token; // for Spring Boot back-end
+      config.headers = {
+        ...config.headers,
+        ["x-access-token"]: token,
+      }; // for Node.js Express back-end
```
Comment (VitaliiDenysiuk) [inline src/services/api.ts]:
```suggestion
      if (config.headers) {
        config.headers["x-access-token"] = token; // for Node.js Express back-end
      }
```
better such way

File: src/services/api.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  }
+);
+
+instance.interceptors.response.use(
+  (res) => {
+    return res;
+  },
+  async (err) => {
+    const originalConfig = err.config;
+    if (originalConfig.url !== "/auth/login" && err.response) {
```
Comment (VitaliiDenysiuk) [inline src/services/api.ts]:
```suggestion
    if (originalConfig.url !== "/api/auth/login" && err.response) {
```

File: src/services/api.ts [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+const instance = axios.create({
+  baseURL: "http://localhost:3001/api", // /auth/login[,register]
+  headers: {
+    "Content-Type": "application/json",
+  },
+});
+
+function getRefreshToken() {
+  return instance.post("/auth/refreshtoken", {
```
Comment (VitaliiDenysiuk) [inline src/services/api.ts]:
```suggestion
  return instance.post("/api/auth/refreshtoken", {
```

File: src/services/api.ts [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        }
+      }
+    }
+    return Promise.reject(err);
+  }
+);
+
+export const registerUser = async () => {
+  try {
+    const res = await instance.post("/auth/register", {
```
Comment (VitaliiDenysiuk) [inline src/services/api.ts]:
```suggestion
    const res = await instance.post("/api/auth/register", {
```

File: src/services/api.ts [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    localStorage.setItem("accessToken", accessToken);
+    localStorage.setItem("refreshToken", refreshToken);
+  } catch (err) {
+    console.warn(err);
+  }
+};
+
+export const testAuthorized = async () => {
+  try {
+    const res = await instance.get("/testAuthorized");
```
Comment (VitaliiDenysiuk) [inline src/services/api.ts]:
```suggestion
    const res = await instance.get("/api/testAuthorized");
```

**Review by VitaliiDenysiuk — CHANGES_REQUESTED** [review]

**Review by VladislavPokhyl — APPROVED** [review]

File: src/services/api.ts (hunk 1 above)
Comment (KirillSerg) [reply src/services/api.ts:20]: thanks for knowledge & link.
Done.

File: src/services/api.ts (hunk 2 above)
Comment (KirillSerg) [reply src/services/api.ts]: thanks. Done

File: src/services/api.ts (hunk 3 above)
Comment (KirillSerg) [reply src/services/api.ts]: done

File: src/services/api.ts (hunk 4 above)
Comment (KirillSerg) [reply src/services/api.ts]: done

File: src/services/api.ts (hunk 5 above)
Comment (KirillSerg) [reply src/services/api.ts]: done

File: src/services/api.ts (hunk 6 above)
Comment (KirillSerg) [reply src/services/api.ts]: done

**Review by VitaliiDenysiuk — DISMISSED** [review]
looks good to me

**Review by VitaliiDenysiuk — APPROVED** [review]
It works. I tested it with the backend.
looks good to me

**Review by alexandrtovmach — CHANGES_REQUESTED** [review]
looks good but not ideal
the current flow refreshing token after failed attempt to get required data, but how about to avoid unnecessary request? You can parse token on frontend side before sending original request to check if it's fresh. Refresh token if it's not fresh and send original request if everything fine.

**Review by alexandrtovmach — APPROVED** [review]

**Pull request merged**